src/order/succ_pred/limit.lean

```diff
@@ -0,0 +1,222 @@
+def is_succ_limit (a : α) : Prop := ∀ b, succ b = a → is_max b
+
+lemma is_succ_limit.is_max (h : is_succ_limit (succ a)) : is_max a := h a rfl
+
+lemma _root_.is_min.is_succ_limit (h : is_min a) : is_succ_limit a :=
+by { rintros b rfl, exact max_of_succ_le (h $ le_succ b) }
+
+lemma is_succ_limit_of_succ_ne (h : ∀ b, succ b ≠ a) : is_succ_limit a :=
+by { rintros b rfl, exact (h b).irrefl.elim }
+
+lemma mem_range_succ_of_not_is_succ_limit (h : ¬ is_succ_limit a) : a ∈ range (@succ α _ _) :=
+by { dsimp [range], contrapose! h, exact is_succ_limit_of_succ_ne h }
+
+/-- A value can be built by building it on successors and successor limits.
+
+Note that you need a partial order without a maximum for data built using this to behave nicely on
+successors. -/
+@[elab_as_eliminator] noncomputable def is_succ_limit_rec_on {C : α → Sort*} (b)
+  (hs : Π a, C (succ a)) (hl : Π a, is_succ_limit a → C a) : C b :=
+begin
+  by_cases hb : is_succ_limit b,
+  { exact hl b hb },
+  { rw ←classical.some_spec (mem_range_succ_of_not_is_succ_limit hb),
+    apply hs }
+end
+
+@[simp] theorem is_succ_limit_rec_on_limit {C : α → Sort*} (hs : Π a, C (succ a))
+  (hl : Π a, is_succ_limit a → C a) (hb : is_succ_limit b) :
+  @is_succ_limit_rec_on α _ _ C b hs hl = hl b hb :=
+by { classical, exact dif_pos hb }
+
+section no_max_order
+variables [no_max_order α]
+
+lemma is_succ_limit.succ_ne (h : is_succ_limit a) (b) : succ b ≠ a := λ hb, not_is_max b (h b hb)
+
+@[simp] lemma not_is_succ_limit_succ (a : α) : ¬ is_succ_limit (succ a) := λ h, h.succ_ne _ rfl
+
+lemma is_succ_limit_iff_succ_ne : is_succ_limit a ↔ ∀ b, succ b ≠ a :=
+⟨is_succ_limit.succ_ne, is_succ_limit_of_succ_ne⟩
+
+lemma is_succ_limit_of_succ_lt (H : ∀ a < b, succ a < b) : is_succ_limit b :=
+by { rintros a rfl, exact (H a (lt_succ a)).false.elim }
+
+lemma not_is_succ_limit_iff : ¬ is_succ_limit a ↔ a ∈ range (@succ α _ _) :=
+by { simp_rw [is_succ_limit_iff_succ_ne, not_forall, not_ne_iff], refl }
+
+end no_max_order
+
+section order_bot
+variable [order_bot α]
+
+lemma is_succ_limit_bot : is_succ_limit (⊥ : α) := is_min_bot.is_succ_limit
+
+end order_bot
+
+section is_succ_archimedean
+variable [is_succ_archimedean α]
+
+lemma is_succ_limit.is_min [no_max_order α] (h : is_succ_limit a) : is_min a :=
+λ a ha, begin
+  rcases ha.exists_succ_iterate with ⟨_ | n, rfl⟩,
+  { exact le_rfl },
+  { rw iterate_succ_apply' at h,
+    exact (not_is_succ_limit_succ _ h).elim }
+end
+
+@[simp] lemma is_succ_limit_iff [no_max_order α] : is_succ_limit a ↔ is_min a :=
+⟨is_succ_limit.is_min, is_min.is_succ_limit⟩
+
+lemma not_is_succ_limit [no_min_order α] [no_max_order α] : ¬ is_succ_limit a :=
+by simp
+
+end is_succ_archimedean
+end preorder
+
+section partial_order
+variables [partial_order α] [succ_order α]
+
+section no_max_order
+variables [no_max_order α]
+
+lemma is_succ_limit.succ_lt (hb : is_succ_limit b) (ha : a < b) : succ a < b :=
+by { rw [lt_iff_le_and_ne, succ_le_iff], exact ⟨ha, hb.succ_ne a⟩ }
```

Comment:
That doesn't need `no_max_order α`. Also, doesn't it work in preorders?

Reply:
I don't think this works on preorders, you could have `succ a` and `b` be order-indistinguishable from one another.

Reply:
I guess that means the right definition of `is_succ_limit` is `∀ b, antisymm_rel (succ b) a → is_max b`! But yeah we don't care about preorders enough to make this work.

Reply:
I mean, I could make it work, and it might make some of the typeclass assumptions make more sense. After all, this is one of those definitions where you're supposed to access the API rather than the bare definition.

Reply:
Hmm... actually, doing this ends up making a bunch of other theorems, namely those that involve `succ a ≠ b` for `b` a successor limit, need partial orders now. So I think it's best to keep the definition as is. Not like anyone will try using this on preorders, hopefully.